### PR TITLE
Multi stage builds in Rahti

### PR DESCRIPTION
**Pull request by lvarin**

### CODE REVIEW

**Comment by lvarin** [comment]
https://csc-guide-preview.rahtiapp.fi/origin/multi-stage/cloud/rahti/tutorials/multi-stage-builds/

**Comment by attesillanpaa** [comment]
Hi! Would you add this (and maybe others if missing) tutorial also in this list: https://docs.csc.fi/support/tutorials/#rahti-container-cloud (and assign a reviewer?)

**Comment by lvarin** [comment]
Yes, I will do it

File: docs/cloud/rahti/tutorials/multi-stage-builds.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# Multistage builds
+
+The idea behind _multistage builds_ is to make easy to build smaller container images by facilitating excluding the intermediate build files from the final product.
+
+Smaller container images take less space in disk, this means they take less time to download and deploy. During the compilation of software, it is common to need, a compiler, several library dependencies, and intermediate objects that will not be needed during the execution of the program. Multistage builds allow you to define two or more docker build stages in the same `Dockerfile`. They will be executed in order one after the other, and each "stage" will be able to copy files from the previous ones. This way, we can easily and in the same build process, build the software and then use only the files that we actually need for the execution.
```
Comment (tdeneke) [inline docs/cloud/rahti/tutorials/multi-stage-builds.md:5]:
```suggestion
Smaller container images take less space on disk, meaning they take less time to download and deploy. During the compilation of software, it is common to need a compiler, several library dependencies, and intermediate objects that will not be needed during the execution of the program. Multistage builds allow you to define two or more docker build stages in the same `Dockerfile`. They will be executed in order one after the other, and each "stage" will be able to copy files from the previous ones. This way, we can easily and in the same build process, build the software and then keep only the files that we actually need for the execution.
```

File: docs/cloud/rahti/tutorials/multi-stage-builds.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        fmt.Fprintf(w, uuid)
+    })
+
+    fmt.Print("Starting server in port 8080...\n")
+
+    http.ListenAndServe(":8080", nil)
+}
+```
+
+The `dockerfile` can be divided in two parts (or stages), each starting by the `FORM` instruction:
```
Comment (tdeneke) [inline docs/cloud/rahti/tutorials/multi-stage-builds.md:57]:
```suggestion
The `dockerfile` can be divided into two parts (or stages), each starting by the `FORM` instruction:
```

File: docs/cloud/rahti/tutorials/multi-stage-builds.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+This will produce the image called `go-server:latest`. To check the size of the image just run:
+
+```sh
+$ docker images go-server
+REPOSITORY   TAG       IMAGE ID       CREATED        SIZE
+go-server    latest    514583ab6fa1   17 hours ago   13MB
+```
+
+it should give you approximately 13MB, which more than half (~7MB) is the compiled code.
```
Comment (tdeneke) [inline docs/cloud/rahti/tutorials/multi-stage-builds.md:76]:
```suggestion
it should give you approximately 13MB, which is more than half (~7MB) is the compiled code.
```

File: docs/cloud/rahti/tutorials/multi-stage-builds.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+$ docker images golang:1.12.6-stretch
+REPOSITORY   TAG              IMAGE ID       CREATED       SIZE
+golang       1.12.6-stretch   9fe4cdc1f173   2 years ago   774MB
+```
+
+This same small image (`go-server:latest`) is of course also achievable by other methods. You can build the code outside of docker and then copy it to the `alpine` image. You can mount the code directory into the build image, build it and then again copy the compiled product into the `alpine` image. But none of these methods are as easy and compact as this one.
+
+## Usage in Rahti
+
+In order to test this in Rahti, one only needs to login in Rahti, select the correct project and run:
```
Comment (tdeneke) [inline docs/cloud/rahti/tutorials/multi-stage-builds.md:90]:
```suggestion
In order to test this in Rahti, one only needs to login in Rahti, select the correct project, and run:
```

**Review by tdeneke — COMMENTED** [review]
Content is very good. Only minor comments on wording.

**Review by tdeneke — APPROVED** [review]
looks good!

**Pull request merged**